Pocket.Logger/Core/Formatter.cs

```diff
@@ -1,193 +1,312 @@
+            {
+                stopCaching = true;
             }
+
+            return new Formatter(t);
         }
+    }
 
-        internal class FormatterResult : IReadOnlyList<(string Name, object Value)>
+    internal class FormatterResult : IReadOnlyList<(string Name, object Value)>
+    {
+        private readonly StringBuilder formattedMessage;
+
+        public FormatterResult(StringBuilder formattedMessage)
         {
-            private readonly StringBuilder formattedMessage;
+            this.formattedMessage = formattedMessage;
+        }
 
-            public FormatterResult(StringBuilder formattedMessage)
-            {
-                this.formattedMessage = formattedMessage;
-            }
+        private readonly List<(string Name, object Value)> properties = new();
+
+        public void Add(string key, object value) => properties.Add((key, value));
+
+        public IEnumerator<(string Name, object Value)> GetEnumerator() => properties.GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        public int Count => properties.Count;
+
+        public (string Name, object Value) this[int index] => properties[index];
+
+        public override string ToString() => formattedMessage?.ToString() ?? "";
+    }
+}
+
+internal static partial class Format
+{
+    public static string ToLogString(
+       this in (
+           byte LogLevel,
+           DateTime TimestampUtc,
+           Func<(string Message, (string Name, object Value)[] Properties)> Evaluate,
+           Exception Exception,
+           string OperationName,
+           string Category,
+           (string Id,
+           bool IsStart,
+           bool IsEnd,
+           bool? IsSuccessful,
+           TimeSpan? Duration) Operation) e)
+    {
+        var (message, _) = e.Evaluate();
+
+        var logLevelString =
+            LogLevelString((LogLevel)e.LogLevel,
+                           e.Operation.IsStart,
+                           e.Operation.IsEnd,
+                           e.Operation.IsSuccessful,
+                           e.Operation.Duration);
+
+        return
+            $"{e.TimestampUtc:o} {e.Operation.Id.IfNotEmpty()}{e.Category.IfNotEmpty()}{e.OperationName.IfNotEmpty()} {logLevelString} {message} {e.Exception}";
+    }
+
+    static partial void CustomizeLogString(object value, ref string output);
+
+    internal static string ToLogString(this object value)
+    {
+        string output = null;
+
+        CustomizeLogString(value, ref output);
+
+        if (output is not null)
+        {
+            return output;
+        }
+
+        if (value is IEnumerable enumerable and not string)
+        {
+            return $"[ {string.Join(", ", enumerable.Cast<object>())} ]";
+        }
 
-            private readonly List<(string Name, object Value)> properties = new List<(string, object )>();
+        if (value is null)
+        {
+            return "[null]";
```

Comment:
Consider moving this early out to the top.

Reply:
The goal was to also allow customizing the null case.